p2p/net/swarm/black_hole_detector.go

```diff
@@ -0,0 +1,208 @@
+	}
+
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	b.requests++
+	if b.requests == b.n {
+		b.requests = 0
+	}
+	b.trackMetrics()
+	return b.state, (b.state == blackHoleStateAllowed) || (b.requests == 0)
+}
+
+func (b *blackHoleFilter) reset() {
+	b.successes = 0
+	b.failures = 0
+	b.outcomes = b.outcomes[:0]
+	b.updateState()
+}
+
+func (b *blackHoleFilter) updateState() {
+	st := b.state
+	successFraction := 0.0
+	if len(b.outcomes) < b.n {
+		b.state = blackHoleStateAllowed
+	} else {
+		successFraction = float64(b.successes) / float64(b.successes+b.failures)
+		if successFraction >= b.minSuccessFraction {
+			b.state = blackHoleStateAllowed
+		} else {
+			b.state = blackHoleStateBlocked
+		}
+	}
+	if st != b.state {
+		if b.state == blackHoleStateAllowed {
+			log.Debugf("%s blackHoleDetector state changed to Allowed", b.name)
+		} else {
+			log.Debugf("%s blackHoleDetector state changed to Blocked. Success fraction is %0.3f", b.name, successFraction)
+		}
+	}
+}
+
+func (b *blackHoleFilter) trackMetrics() {
+	if b.metricsTracer == nil {
+		return
+	}
+	successFraction := 0.0
+	if b.successes+b.failures != 0 {
+		successFraction = float64(b.successes) / float64(b.successes+b.failures)
+	}
+	b.metricsTracer.UpdatedBlackHoleFilterState(
+		b.name,
+		b.state,
+		b.n-b.requests,
+		successFraction,
+	)
+}
+
+// blackHoleDetector provides UDP and IPv6 black hole detection using a `blackHoleFilter`
+// for each. For details of the black hole detection logic see `blackHoleFilter`
+type blackHoleDetector struct {
+	udp, ipv6 *blackHoleFilter
+}
+
+func (d *blackHoleDetector) IsAllowed(addr ma.Multiaddr) bool {
+	if !manet.IsPublicAddr(addr) {
+		return true
+	}
+
+	udpState, udpAllowed := blackHoleStateAllowed, true
+	if d.udp != nil && isProtocolAddr(addr, ma.P_UDP) {
+		udpState, udpAllowed = d.udp.IsAllowed()
+	}
+
+	ipv6State, ipv6Allowed := blackHoleStateAllowed, true
+	if d.ipv6 != nil && isProtocolAddr(addr, ma.P_IP6) {
+		ipv6State, ipv6Allowed = d.ipv6.IsAllowed()
+	}
+
+	// Allow all probes irrespective of the state of the other filter
+	if (udpState == blackHoleStateBlocked && udpAllowed) ||
+		(ipv6State == blackHoleStateBlocked && ipv6Allowed) {
+		return true
+	}
+	return (udpAllowed && ipv6Allowed)
+}
+
+// RecordOutcome updates the state of the relevant `blackHoleFilter` for addr
+func (d *blackHoleDetector) RecordOutcome(addr ma.Multiaddr, success bool) {
+	if !manet.IsPublicAddr(addr) {
+		return
+	}
+	if d.udp != nil && isProtocolAddr(addr, ma.P_UDP) {
+		d.udp.RecordOutcome(success)
+	}
+	if d.ipv6 != nil && isProtocolAddr(addr, ma.P_IP6) {
+		d.ipv6.RecordOutcome(success)
+	}
+}
+
+func newBlackHoleDetector(detectUDP, detectIPv6 bool, mt MetricsTracer) *blackHoleDetector {
+	d := &blackHoleDetector{}
+	if detectUDP {
+		d.udp = &blackHoleFilter{n: 100, minSuccessFraction: 0.01, name: "UDP", metricsTracer: mt}
+	}
+	if detectIPv6 {
+		d.ipv6 = &blackHoleFilter{n: 100, minSuccessFraction: 0.01, name: "IPv6", metricsTracer: mt}
```

Comment:
This would lead to problems in the pathological case where peers never listened on QUIC on IPv6, right? Then the QUIC dials on IPv4 would succeed, canceling the TCP dials on IPv6, leading us to conclude that we're blackholed.

I think this is fine, but we should add some justification in a comment.

Reply:
Yes in this case we will be incorrect for a few probes. Another alternative is to allow 3-4 probes which reduces the chances that this would happen.